internal/parallel/tar_ball_file_packer.go

```diff
@@ -0,0 +1,108 @@
+package parallel
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+	"github.com/wal-g/wal-g/internal/ioextensions"
+	"github.com/wal-g/wal-g/internal/limiters"
+	"github.com/wal-g/wal-g/utility"
+	"golang.org/x/sync/errgroup"
+)
+
+type FileNotExistError struct {
+	error
+}
+
+func NewFileNotExistError(path string) FileNotExistError {
+	return FileNotExistError{errors.Errorf(
+		"%s does not exist, probably deleted during the backup creation\n", path)}
+}
+
+func (err FileNotExistError) Error() string {
+	return fmt.Sprintf(tracelog.GetErrorFormatter(), err.error)
+}
+
+type TarBallFilePacker interface {
+	PackFileIntoTar(cfi *ComposeFileInfo, tarBall internal.TarBall) error
+}
+
+type TarBallFilePackerOptions struct {
+	verifyPageChecksums   bool
+	storeAllCorruptBlocks bool
+}
+
+func NewTarBallFilePackerOptions(verifyPageChecksums, storeAllCorruptBlocks bool) TarBallFilePackerOptions {
+	return TarBallFilePackerOptions{
+		verifyPageChecksums:   verifyPageChecksums,
+		storeAllCorruptBlocks: storeAllCorruptBlocks,
+	}
+}
+
+type RegularTarBallFilePacker struct {
+	files BundleFiles
+}
+
+func NewRegularTarBallFilePacker(files BundleFiles, options TarBallFilePackerOptions) *RegularTarBallFilePacker {
+	return &RegularTarBallFilePacker{
+		files: files,
+	}
+}
+
+func (p *RegularTarBallFilePacker) PackFileIntoTar(cfi *ComposeFileInfo, tarBall internal.TarBall) error {
+	fileReadCloser, err := startReadingFile(cfi.Header, cfi.FileInfo, cfi.Path)
+	if err != nil {
+		switch err.(type) {
+		case FileNotExistError:
+			// File was deleted before opening.
+			// We should ignore file here as if it did not exist.
+			tracelog.WarningLogger.Println(err)
+			return nil
+		default:
+			return err
+		}
+	}
+	errorGroup, _ := errgroup.WithContext(context.Background())
```

Comment:
Since there is only one goroutine left, `errorGroup` is no longer needed. You can remove the `errorGroup` and goroutine completely.